### PR TITLE
Kube certs collector davidr

### LINKED ISSUE #762: Collector & Analyzer for certificates
**Describe the rationale for the suggested feature.**

Several support cases have come through where certificates have expired, we would like to be able to explain this at the analyzer stage so folks know immediately what to do about it.

**Describe the feature**

Ensure that the certificate expiry information is available in support bundles for:
* Kubernetes
* weave
* kURL registry
* Envoy

Create an analyzer that will highlight expired certificates from any of the above, along with a URL that will show instructions for remedy.

**Describe alternatives you've considered**

.

**Additional context**

Related Shortcut: https://app.shortcut.com/replicated/story/42738/add-collectors-and-analyzers-for-expired-certificates

Envoy: https://community.replicated.com/t/kurl-how-to-manually-rotate-an-expired-certificate-for-envoy/889

kURL Registry: https://community.replicated.com/t/what-to-do-if-the-kurl-registry-certificates-have-expired/955

Contour: https://community.replicated.com/t/kots-is-it-possible-to-configure-tls-certs-for-contour-in-the-embedded-installation/390/2

Checking the Kubernetes certs themselves is more difficult as it requires `sudo kubeadm certs check-expiration` which requires sudo.

**Definition of done**

- [x] Write a spec for the collectors and analysers - Done [here](https://github.com/replicatedhq/troubleshoot/issues/762#issuecomment-1285916327)
- [x] Develop an in-cluster collector
  - [x] Certs from secrets
  - [x] Certs from config maps
  - [ ] Certs from endpoints
  - [x] In-cluster kubernetes cluster certs - most are in `kube-*` config maps
- [x] Develop a host collector
  - [x] Kubernetes cluster certs from host paths i.e `/etc/kubernetes/pki`, `/var/lib/kubelet/pki` on a kubeadm-based cluster
  - [ ] Certs from endpoints
  - [x] Certs from a list of paths in the host e.g `/etc/myservice/*.crt`
- [x] Develop an analyser - reviews expiry amongst other things
- [ ] The default specs we provide in [troubleshoot-specs](https://github.com/replicatedhq/troubleshoot-specs) including analyzers for the above certificates with links to docs to fix them. Collect the following
  - [ ] Kubernetes (host & in-cluster collectors)
  - [ ] kURL registry
  - [ ] Contour/envoy
  - [ ]  KOTS Admin Console (kotsadm-tls)
- [x] Document both the collector and analyser in [troubleshoot.sh docs](https://troubleshoot.sh/docs/)

### PR DESCRIPTION
**Pull request by drohnow**
Context: Parse K8s SSL certificates so they can be analyzed for validity (ie: expiration date, other)

Summary of Change:
- Added a host collector that collects SSL certificates from the following path host path: /etc/kubernetes/pki
- Output is a json file (see example below)

file: ssl_certificates/kube_ssl_certificates.json

[
	{
		"Certificate Name": "apiserver-etcd-client.crt",
		"DNS Names": null,
		"Issuer": "etcd-ca",
		"Issuer Organizations": null,
		"Certificate Expiration Date": "2023-12-14T20:18:59Z",
		"IsValid": true,
		"Location": {
			"File Path": "/etc/kubernetes/pki/"
		}
	},
	{
		"Certificate Name": "apiserver-kubelet-client.crt",
		"DNS Names": null,
		"Issuer": "kubernetes",
		"Issuer Organizations": null,
		"Certificate Expiration Date": "2023-12-14T20:18:57Z",
		"IsValid": true,
		"Location": {
			"File Path": "/etc/kubernetes/pki/"
		}
	},
	{
		"Certificate Name": "apiserver.crt",
		"DNS Names": [
			"davidr-timeouts",
			"kubernetes",
			"kubernetes.default",
			"kubernetes.default.svc",
			"kubernetes.default.svc.cluster.local"
		],
		"Issuer": "kubernetes",
		"Issuer Organizations": null,
		"Certificate Expiration Date": "2023-12-14T20:18:57Z",
		"IsValid": true,
		"Location": {
			"File Path": "/etc/kubernetes/pki/"
		}
	},

## Checklist

- [ ] New and existing tests pass locally with introduced changes.
	Note tests created yet.
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] The commit message(s) are informative and highlight any breaking changes
- [ ] Any documentation required has been added/updated. For changes to https://troubleshoot.sh/ create a PR [here](https://github.com/replicatedhq/troubleshoot.sh/pulls)

## Does this PR introduce a breaking change?
- [ ] Yes
- [X] No

TODO: add functionality to retrieve and validate ssl certificate via URL.  However, please review progress to date.

### CODE REVIEW

**Comment by diamonwiggins** [comment]
@drohnow I'm curious if we're duplicating the functionality of the existing TLS certificate host collector/analyzer here? See - https://github.com/replicatedhq/troubleshoot-specs/blob/main/host/cluster-down.yaml#L18-L25

File: pkg/collect/host_kube_ssl_cert_collector.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	return getCollectorName(c)
+}
+
+func (c *CollectKubeSSLCertInfo) IsExcluded() (bool, error) {
+	return isExcluded(c.hostCollector.Exclude)
+}
+
+// SSL Certificate Struct
+type KubeSSLCertCollect struct {
+	CertName         string    `json:"Certificate Name"`
```
Comment (banjoh) [inline pkg/collect/host_kube_ssl_cert_collector.go:32]: The project follows a `camelCase` naming convention meaning this would be `certificateName`

**Review by banjoh — CHANGES_REQUESTED** [review]
This is a good start towards building the collector.

* The collector is not generic as specified in the [issue spec](https://github.com/replicatedhq/troubleshoot/issues/762). Its collecting cert metadata from /etc/kubernetes/pki/*.crt files only. The source of where to fetch the certificates from needs to be driven by config input.
* I suggest we remove SSL references from the collector. Certs can have other purposes than SSL/TLS e.g authentication, digital signatures

Perhaps we can talk about this; I believe it would be important to surface all k8s related certificates holistically to validate certificate status health.

File: pkg/collect/host_kube_ssl_cert_collector.go (hunk 1 above)
Comment (drohnow) [reply pkg/collect/host_kube_ssl_cert_collector.go:32]: ack

**Comment by drohnow** [comment]
Closing PR; stale.

**Pull request closed without merging**